Fusion & Construction file changes (to reflect the fusion changes)

sjjs7 opened the pull request:
Will let you know once these have passed holtest!

sjjs7 committed:
Update fusion

sjjs7 committed:
Update ConstructionTactics.ml

sjjs7 committed:
Update Induction_Schema_Example.ml

sjjs7 committed:
Update LEM_Example.ml

sjjs7 committed:
Update QuotationOperatorTests.ml

sjjs7 committed:
Update QuotationTactics.ml

JacquesCarette commented on Constructions/Induction_Schema_Example.ml:
```diff
@@ -1,16 +1,19 @@
 
 (*Trivial proof that x is effecitve in x + 3*)
-prove(effectiveIn `x:num` `x + 3`,
+prove(mk_neg (mk_not_effective_in `x:num` `x + 3` `y:num`),
```
Why the change from a positive statement to a double-negative?

sjjs7 replied:
Yes I guess that does look off- it's because I changed the default from effective-in to not-effective-in. I should still have an effective-in statement though, will add.

JacquesCarette commented on Constructions/QuotationOperatorTests.ml:
```diff
@@ -1,22 +1,24 @@
-(*Similair to the tests for epsilon, this will prove a few theoremms to verify that _Q_ and Q_ _Q work as intended, along with a few tests with OCaml functions to ensure the Quote term is working correctly*)
+needs "Constructions/QuotationTactics.ml";;
+needs "Constructions/epsilon.ml";;
+(*Similair to the tests for epsilon, this will prove a few theorems to verify that _Q_ and Q_ _Q work as intended, along with a few tests with OCaml functions to ensure the Quote term is working correctly*)
```
You might want to fix the typo in the comment while you're at it...

JacquesCarette commented on fusion.ml:
```diff
@@ -75,6 +75,7 @@ module type Hol_kernel =
       val variant : term list -> term -> term
       val vsubst : (term * term) list -> term -> term
       val qsubst : (term * term) list -> term -> term
+      val BETA_RED_BY_SUB : term -> thm 
```
This stands out (all caps) amongst things that aren't. Does it belong in this part of the signature?

JacquesCarette commented on fusion.ml:
```diff
@@ -231,6 +230,8 @@ module Hol : Hol_kernel = struct
           (Tyapp(_,args)) -> itlist (union o tyvars) args []
         | (Tyvar v as tv) -> [tv]
 
+  let is_defined_type ty = if tyvars ty = [] then true else false 
```
Argh!  Anti-pattern!  This should be
```ocaml
let is_defined_type ty = ityvars ty = []
```

sjjs7 replied:
Yes that's awful!

JacquesCarette commented on fusion.ml:
```diff
@@ -439,6 +442,23 @@ let rec type_subst i ty =
+(* ------------------------------------------------------------------------- *)
+
+  let addThm tm = match tm with 
+      | Sequent(asl,c) -> if (forall (fun a -> is_proven_thm a) asl) then 
```
I much prefer to put the nested if on the next line, and indent it 4 from the `|` (which is 2 from the `Sequent`). That way the code isn't so wide, with lots of wasted space on the left.

JacquesCarette commented on fusion.ml:
```diff
@@ -673,29 +681,36 @@ let rec type_subst i ty =
                         | _ -> false) theta
       then vsubst theta else failwith "vsubst: Bad substitution list"
 
+  let BETA_RED_BY_SUB tm = if is_comb tm && is_abs(fst(dest_comb tm))
```
Same here: the `if` should be on the next line, and indented just 2.

JacquesCarette commented on fusion.ml:
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
-      | _ -> failwith "TERM_TO_CONSTRUCTION"
-  
-  let CONSTRUCTION_TO_TERM tm = try Sequent([],safe_mk_eq tm (mk_quote (constructionToTerm tm))) with Failure _ -> failwith "CONSTRUCTION_TO_TERM"
+  let is_proper_construc c = if can constructionToTerm c then true else false 
```
anti-pattern

JacquesCarette commented on fusion.ml:
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
+  let LAW_OF_QUO tm = match tm with
+    | Var(_,_) | Const(_,_) | Comb(_,_) | Abs(_,_) | Quote(_,_) | Hole(_,_) -> 
+      if is_eval_free tm then Sequent([],safe_mk_eq (mk_quote tm) (termToConstruction tm))
+      else failwith "Not eval free"
```
This is a bad error message. You should be able to print the occurrence of `Eval` that causes the problem. This means that `is_eval_free` shouldn't produce a boolean, but rather a data-type with 2 constructors (call them `Ok` and `Problem` where `Problem` contains the payload necessary for you to print a decent error.

JacquesCarette commented on fusion.ml:
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
+    | Var(_,_) | Const(_,_) | Comb(_,_) | Abs(_,_) | Quote(_,_) | Hole(_,_) -> 
+      if is_eval_free tm then Sequent([],safe_mk_eq (mk_quote tm) (termToConstruction tm))
+      else failwith "Not eval free"
+    | Eval(_,_) -> failwith "Not eval free"
```
Here, giving a better error is easy.

JacquesCarette commented on fusion.ml:
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
+      if is_eval_free tm then Sequent([],safe_mk_eq (mk_quote tm) (termToConstruction tm))
+      else failwith "Not eval free"
+    | Eval(_,_) -> failwith "Not eval free"
+    | _ -> failwith "Not a proper term"  
```
anti-pattern: catch-all match.

JacquesCarette commented on fusion.ml:
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
-    | _ -> failwith "TERM_TO_CONSTRUCTION_CONV"
+  let LAW_OF_QUO_CONV tm = LAW_OF_QUO tm
+
+  let rec QUOTE_TO_CONSTRUCTION_CONV tm = match tm with
```
you can use `function` for this pattern, i.e.
```ocaml
let foo xx = match xx with
```
is just
```ocaml
let foo = function
```
(there are more places where this can be used)

JacquesCarette commented on fusion.ml:
```diff
@@ -1212,75 +1244,73 @@ let rec type_subst i ty =
+      | _ -> failwith "LAW_OF_DISQUO: Not evaluating a quotation!"
 
+  let rec LAW_OF_DISQUO_CONV tm = match tm with
+    | Comb(a,b) -> (try (LAW_OF_DISQUO_CONV a) with Failure _ -> try (LAW_OF_DISQUO_CONV b) with Failure _ -> failwith "LAW_OF_DISQUO_CONV")
```
Same comments about error messages.

JacquesCarette commented on fusion.ml:
```diff
@@ -1212,75 +1244,73 @@ let rec type_subst i ty =
+                        Sequent([],(internal_make_imp iet (safe_mk_eq (Eval(Comb(Const("Quo",makeHolFunction (makeHolType "epsilon" []) (makeHolType "epsilon" [])),tm),Tyapp("epsilon",[]))) tm )))
+                      else failwith "QUO_DISQUO"
+
+  let ABS_DISQUO var tm = if not (is_var var) then failwith "ABS_DISQUO: not a variable" 
```
I have a feeling that this could be written more clearly with pattern matching than via a whole bunch of if-then.

sjjs7 committed:
Update QuotationOperatorTests.ml

sjjs7 committed:
Update Induction_Schema_Example.ml

sjjs7 committed:
Update fusion.ml

Biggest change was the addition of an exception (Problem) and new type (check) which is to help throw exceptions.

sjjs7 commented:
Ok I think I fixed the issues, let me know what you think. 
(I still have to run holtest)

sjjs7 committed:
Update simp.ml

Change is_eval_free to eval_free to reflect change in fusion

JacquesCarette commented on fusion.ml:
```diff
@@ -313,49 +327,48 @@ let rec type_subst i ty =
+  let rec is_eval_free tm = match tm with 
+    | Var(_,_) -> Ok
+    | Const(_,_) -> Ok
+    | Comb(a,b) -> if is_eval_free a = Ok then is_eval_free b else is_eval_free a 
```
You want to let-bind `is_eval_free a` so that you can re-use it in the `else` clause.  [If Ocaml had monads, it would be even simpler, but...]

JacquesCarette commented on fusion.ml:
```diff
@@ -417,11 +431,14 @@ let rec type_subst i ty =
+  let mk_quote t = 
+     match (is_eval_free t) with 
+        | Ok -> Quote(t,qcheck_type_of t)
+        | Issue tm -> raise (Problem("Cannot quote terms with evaluations. This term contains one:", tm))
```
Definitely better.

sjjs7 committed:
Update fusion.ml

Changes to fusion as requested by Bill.
Remove old code, fix comments & layout of inference rules.

sjjs7 commented:
I guess I better check these changes as well...will let you know.

sjjs7 committed:
Update fusion.ml

Fix error!!

sjjs7 commented:
Alright, it's fine, we're good to go.
